cachi2/core/package_managers/pip.py

```diff
@@ -1289,13 +1287,13 @@ def from_line(cls, line: str, options: list[str]) -> Optional["PipRequirement"]:
         :return: PipRequirement instance
         """
         to_be_parsed = line
-        qualifiers = {}
+        qualifiers: dict[str, str] = {}
         requirement = cls()
 
-        direct_access_kind, is_direct_access = cls._assess_direct_access_requirement(line)
-        if is_direct_access:
+        direct_access_kind = cls._assess_direct_access_requirement(line)
+        if direct_access_kind:
             if direct_access_kind in ["url", "vcs"]:
-                requirement.kind = direct_access_kind
+                requirement.kind = cast(Literal["pypi", "url", "vcs"], direct_access_kind)
```

Comment:
A non-blocking question: why is this cast necessary?

Reply:
mypy refuses to accept `requirement.kind = direct_access_kind` where `direct_access_kind` is type str.

Reply:
This feels like `_assess_direct_access_requirement` is mistyped. If it returned `Optional[Literal[...]` instead of a string then mypy would not have complained. And since it effectively works as a constructor for this type it would be fit to reject an incorrect input value there.

Reply:
> If it returned Optional[Literal[...]

It did, and mypy made the same complaint. Also, returning None here is redundant and type-confusing. Using `""` as False means we can keep it string-typed (also very Pythonic).

> reject an incorrect input value there

See above re: empty string

Reply:
> It did, and mypy made the same complaint.
Did it make it on the same line though? For me mypy started complaining about return type of `_assess_direct_access_requirement` instead of the line it was used on. Now if someone is to use a cast then I'd say that the return statement is a better place for that. The problem here is that we want to simultaneously use a wide type and also have it refined. If this is important I would suggest introducing an enum just for that. It would also cleanly handle validation (which is now pushed to L1302).

Using `Optional` makes this an `Option` or `Maybe` type (depending on once's preference for proper statically typed languages), which is arguably a cleaner way to convey intent from type design perspective. This is probably not too Pythonic, OTOH one could argue that having casts in code is even less Pythonic.

Reply:
> one could argue that having casts in code is even less Pythonic

Absolutely agree there - there's a lot about mypy that is not very Pythonic

Reply:
> Did it make it on the same line though?

Yes

Reply:
Oh, I know what is missing: currently the helper returns unrecognized kind as is and lets its client fail. To make it cleaner we could either add an "unsupported" literal, and have a cast in return, or better yet raise right in the helper. Or have an actual type to represent kinds. Anyway, this is not a blocker, and probably belongs to a separate refactoring effort. The change looks good to me  (except the `cast`, casts will never look good to me, sorry casts) and I have +1-ed it.

Reply:
> currently the helper returns unrecognized kind as is and lets its client fail. To make it cleaner we could ... raise right in the helper

Nice! :taco:

Reply:
walrus++